File: android/app/src/sideload/java/org/getlantern/lantern/activity/UpdateActivity.java
package org.getlantern.lantern.activity;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.text.Html;
import android.view.WindowManager;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.core.content.FileProvider;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.Click;
import org.androidannotations.annotations.EActivity;
import org.androidannotations.annotations.Extra;
import org.androidannotations.annotations.ViewById;
import org.getlantern.lantern.BuildConfig;
import org.getlantern.lantern.R;
import org.getlantern.lantern.model.Utils;
import org.getlantern.lantern.util.ApkSignatureVerifier;
import org.getlantern.lantern.util.SignatureVerificationException;
import org.getlantern.mobilesdk.Logger;

import java.io.File;

import internalsdk.Internalsdk;
import internalsdk.Updater;

@EActivity(R.layout.activity_updater)
public class UpdateActivity extends Activity implements ActivityCompat.OnRequestPermissionsResultCallback {
    private static final String TAG = UpdateActivity.class.getName();
    private static final int REQUEST_CODE_REQUEST_INSTALL_PACKAGES = 1252;

    private UpdaterTask updaterTask;
    private ProgressDialog progressBar;
    private boolean fileDownloading = false;

    static boolean active = false;

    @Extra("updateUrl")
    String updateUrl;

    @ViewById
    TextView updateAvailable;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // if not in dev mode, prevent screenshots of this activity by other apps
        if (!BuildConfig.DEVELOPMENT_MODE) {
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_SECURE);
        }
    }

    @AfterViews
    void afterViews() {
        String appName = getResources().getString(R.string.app_name);
        String message = String.format(getResources().getString(R.string.update_available), appName);
        updateAvailable.setText(message);
    }

    @Override
    protected void onStart() {
        super.onStart();
        active = true;
    }

    @Override
    protected void onStop() {
        super.onStop();
        active = false;
    }

    @Click(R.id.notNow)
    void notNowClicked() {
        finish();
    }

    @Click(R.id.installUpdate)
    void installUpdateClicked() {

        Logger.debug(TAG, "Install Update clicked");

        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP_MR1) {

            final String[] permissions = {
                    android.Manifest.permission.REQUEST_INSTALL_PACKAGES
            };

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                if (!getPackageManager().canRequestPackageInstalls()) {
                    startActivityForResult(new Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES)
                                    .setData(Uri.parse(String.format("package:%s", getPackageName()))),
                            REQUEST_CODE_REQUEST_INSTALL_PACKAGES);
                    return;
                }

            } else {

                for (String permission : permissions) {
                    if (ContextCompat.checkSelfPermission(getApplicationContext(), permission) != PackageManager.PERMISSION_GRANTED) {
                        Logger.debug(TAG, "Requesting permission %1$s", permission);
                        // Android is smart enough to only prompt users the lacked permissions
                        ActivityCompat.requestPermissions(this, permissions, 1);
                        return;
                    }
                }
            }
        }

        installUpdate();
    }

    private void installUpdate() {
        Logger.debug(TAG, "Installing update");
        this.fileDownloading = true;

        final String[] updaterParams = {updateUrl};
        updaterTask = new UpdaterTask(this);
        updaterTask.execute(updaterParams);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_CODE_REQUEST_INSTALL_PACKAGES && resultCode == Activity.RESULT_OK) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                if (getPackageManager().canRequestPackageInstalls()) {
                    installUpdate();
                }
            }
        } else {
            // user didn't give the Install Apps From Unknown Sources permission
            // show error message or just do nothing
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            Logger.v(TAG, "Permission: " + permissions[0] + "was " + grantResults[0]);
            installUpdate();
        }
    }

    private class UpdaterTask extends AsyncTask<String, Long, Boolean> implements DialogInterface.OnClickListener {

        private final UpdateActivity activity;
        private final Context context;
        private final File apkDir;
        private final File apkPath;

        public UpdaterTask(final UpdateActivity activity) {
            context = activity.getApplicationContext();
            apkDir = new File(context.getCacheDir(), "updates");
            apkPath = new File(apkDir, "Lantern.apk");
            this.activity = activity;
        }

        @Override
        public void onClick(DialogInterface dialog, int which) {
            //Cancel download task
            fileDownloading = false;
            progressBar.dismiss();
            activity.finish();
        }

        @Override
        protected void onPreExecute() {
            super.onPreExecute();

            progressBar = new ProgressDialog(activity);
            String appName = getString(R.string.app_name);
            progressBar.setMessage(String.format(getString(R.string.updating_lantern), appName));
            progressBar.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);
            progressBar.setIndeterminate(false);
            progressBar.setCancelable(true);
            progressBar.setProgress(0);

            String cancel = getResources().getString(R.string.cancel);

            progressBar.setButton(ProgressDialog.BUTTON_NEGATIVE, cancel, this);
            progressBar.show();
        }

        @Override
        protected Boolean doInBackground(String... params) {

            String updateUrl = params[0];

            Logger.debug(TAG, "Attempting to download update from " + updateUrl);

            try {

                Updater updater = new Updater() {
                    public void progress(long percentage) {
                        publishProgress(percentage);
                    }
                };

                apkDir.mkdirs();
                Internalsdk.downloadUpdate(updateUrl,
                        apkPath.getAbsolutePath(), updater);

                return true;

            } catch (Exception e) {
                Logger.debug(TAG, "Error downloading update: " + e.getMessage());
            }
            return false;
        }

        // show an alert when the update fails
        // and mention where the user can download the latest version
        // this also dismisses the current updater activity
        private void displayError() {
            Utils.showAlertDialog(
                    activity,
                    context.getString(R.string.error_update),
                    context.getString(R.string.tampered_apk),
                    true);
        }

        // show an alert notifying the user that the downloaded apk has been tampered
        private void displayTamperedApk() {
            Utils.showAlertDialog(
                    activity,
                    context.getString(R.string.error_install_update),
                    manualUpdateHTML(),
                    true);
        }

        /**
         * Updating progress bar
         */
        @Override
        protected void onProgressUpdate(Long... progress) {
            super.onProgressUpdate(progress);
            // setting progress percentage
            if (progress[0] != null) {
                progressBar.setProgress(progress[0].intValue());
            }
        }

        // begin the installation by opening the resulting file
        @Override
        protected void onPostExecute(Boolean result) {
            super.onPostExecute(result);

            progressBar.dismiss();

            // update cancelled by the user
            if (!fileDownloading) {
                finish();
                return;
            }

            if (!result) {
                Logger.debug(TAG, "Error trying to install Lantern update");
                Utils.showAlertDialog(this.activity, context.getString(R.string.error_update), manualUpdateHTML(), false);
                return;
            }

            Logger.debug(TAG, "About to install new version of Lantern Android");
            if (!apkPath.isFile()) {
                Logger.error(TAG, "Error loading APK; not found at " + apkPath);
                Utils.showAlertDialog(this.activity, context.getString(R.string.error_update), manualUpdateHTML(), false);
                return;
            }
            try {
                ApkSignatureVerifier.verify(
                        context,
                        apkPath,
                        BuildConfig.SIGNING_CERTIFICATE_SHA256);
            } catch (SignatureVerificationException sfe) {
                Logger.error(TAG, "Error installing update", sfe);
                displayTamperedApk();
                return;
            }
            Intent i = new Intent();
            i.setAction(Intent.ACTION_VIEW);
            i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            Uri apkURI = Uri.fromFile(apkPath);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                i.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                apkURI = FileProvider.getUriForFile(
                        this.context,
                        "org.getlantern.lantern.fileProvider",
                        apkPath);
            }
            i.setDataAndType(apkURI, "application/vnd.android.package-archive");
            this.context.startActivity(i);
            activity.finish();
        }

        private CharSequence manualUpdateHTML() {
            return Html.fromHtml("<span>" + context.getString(R.string.manual_update) + "</span>");
        }
    }
}

File: android/app/src/sideload/kotlin/org/getlantern/lantern/activity/UpdateActivity.kt
package org.getlantern.lantern.activity

import android.os.Bundle
import android.widget.Button
import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.provider.Settings
import android.text.Html
import android.view.View
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.core.app.ActivityCompat
import androidx.lifecycle.lifecycleScope
import internalsdk.Internalsdk
import internalsdk.Updater
import kotlinx.coroutines.*
import kotlinx.coroutines.Dispatchers.IO
import org.getlantern.lantern.BuildConfig
import org.getlantern.lantern.model.Utils
import org.getlantern.lantern.R
import org.getlantern.lantern.util.ApkInstaller
import org.getlantern.lantern.util.ApkSignatureVerifier
import org.getlantern.lantern.util.DeviceInfo
import org.getlantern.lantern.util.SignatureVerificationException
import org.getlantern.mobilesdk.Logger
import java.io.File

open class UpdateActivity : BaseFragmentActivity(), DialogInterface.OnClickListener {

    companion object {
        private val TAG = UpdateActivity::class.java.name
        private const val REQUEST_CODE_PERMISSION = 1252
    }

    private var apkInstaller: ApkInstaller? = null

    lateinit var title: TextView
    lateinit var subTitle: TextView
    lateinit var progressBarLayout: LinearLayout
    lateinit var progressBar: ProgressBar
    lateinit var updateButtons: RelativeLayout
    lateinit var percentage: TextView
    lateinit var installUpdate: Button
    lateinit var notNow: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.update_dialog)
        title = findViewById(R.id.title)
        subTitle = findViewById(R.id.subTitle)
        progressBarLayout = findViewById(R.id.progressBarLayout)
        progressBar = findViewById(R.id.progressBar)
        updateButtons = findViewById(R.id.updateButtons)
        percentage = findViewById(R.id.percentage)
        installUpdate = findViewById(R.id.installUpdate)
        notNow = findViewById(R.id.notNow)
        installUpdate.setOnClickListener {
            installUpdateClicked()
        }
        notNow.setOnClickListener {
            finish()
        }
        subTitle.setText(getString(R.string.update_available, getString(R.string.app_name)))
    }

    fun publishProgress(percent: Long) {
        progressBar.progress = percent.toInt()
        runOnUiThread {
            percentage.text = "$percent%"
        }
    }

    fun showErrorAlert() {
        Utils.showAlertDialog(this, getString(R.string.error_update), manualUpdateHTML(), false)
    }

    // downloadUpdate creates a new instance of Updater and downloads an update via Go
    // Once the download is complete, ApkSignatureVerifier verifies the APK signature
    private fun downloadUpdate(context: Context, apkDir: File, apkPath: File): Boolean {
        val updater = Updater { percent: Long -> publishProgress(percent) }
        try {
            apkDir.mkdirs()
            val result: Boolean = Internalsdk.downloadUpdate(
                DeviceInfo,
                intent.getStringExtra("updateUrl").toString(),
                apkPath.getAbsolutePath(),
                updater,
            )
            if (!result) {
                // show an alert if there was an error downloading the update
                showErrorAlert()
                return false
            }
            if (!apkPath.isFile()) {
                Logger.error(TAG, "Error loading APK; not found at " + apkPath)
                showErrorAlert()
                return false
            }
            ApkSignatureVerifier.verify(
                context,
                apkPath,
                BuildConfig.SIGNING_CERTIFICATE_SHA256,
            )
            return true
        } catch (sfe: SignatureVerificationException) {
            Logger.debug(TAG, "Error verifying update: " + sfe.message)
            displayTamperedApk(context)
        }
        return false
    }

    override fun onDestroy() {
        super.onDestroy()
        apkInstaller?.let { it.unregisterCallback() }
    }

    // show an alert notifying the user that the downloaded apk has been tampered with
    private fun displayTamperedApk(context: Context) {
        Utils.showAlertDialog(
            this,
            context.getString(R.string.error_install_update),
            manualUpdateHTML(),
            true,
        )
    }

    override fun onClick(dialog: DialogInterface, which: Int) {
        finish()
    }

    // installUpdate launches a new coroutine to download an update and serves the APK
    // for installation once the download completes
    private fun installUpdate() {
        val callingActivity = this
        val appName = getString(R.string.app_name)
        // make progress bar layout in update prompt visible and hide the update buttons
        progressBarLayout.setVisibility(View.VISIBLE)
        subTitle.setVisibility(View.GONE)
        subTitle.setText(getString(R.string.update_available, getString(R.string.app_name)))
        updateButtons.setVisibility(View.GONE)
        title.text = String.format(getString(R.string.updating_lantern), appName)
        var context: Context = applicationContext
        var apkDir: File = File(context.cacheDir, "updates")
        val apkPath = File(apkDir, "Lantern.apk")
        apkInstaller = ApkInstaller(this, apkPath)
        lifecycleScope.launch(IO) {
            val success = downloadUpdate(context, apkDir, apkPath)
            if (success) {
                apkInstaller?.execute()
            }
        }
    }

    private fun manualUpdateHTML(): CharSequence {
        return Html.fromHtml("<span>" + getString(R.string.manual_update) + "</span>")
    }

    private fun hasInstallPackagesPermission(): Boolean {
        return PackageManager.PERMISSION_GRANTED ==
                packageManager.checkPermission(Manifest.permission.REQUEST_INSTALL_PACKAGES, packageName)
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    // start the process of installing the update on Android Oreo (8.x)
    private fun startInstallO() {
        val isGranted = packageManager.canRequestPackageInstalls()
        if (isGranted) {
            installUpdate()
        } else {
            val uri: Uri = Uri.parse("package:%s".format(packageName))
            startActivityForResult(
                Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES).setData(uri),
                REQUEST_CODE_PERMISSION,
            )
        }
    }

    fun installUpdateClicked() {
        when {
            Build.VERSION.SDK_INT <= Build.VERSION_CODES.LOLLIPOP_MR1 -> installUpdate()
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.O -> startInstallO()
            !hasInstallPackagesPermission() -> ActivityCompat.requestPermissions(this, arrayOf<String>(Manifest.permission.INSTALL_PACKAGES), 1)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
        super.onActivityResult(requestCode, resultCode, intent)
        if (resultCode != Activity.RESULT_OK) {
            return
        }
        when (requestCode) {
            REQUEST_CODE_PERMISSION -> {
                installUpdate()
            }
        }
    }
}